### ISSUE TITLE
MultiGPU Evaluation Problem

### ISSUE THREAD

**Issue by taoari** [apache/incubator-mxnet#4230]
**The evaluation accuracies do not match when resuming from a previous checkpoint with multi-gpus training. However, this problem does not occur in a single gpu case.**

For example, I am training ResNet-8 on Cifar10 dataset. For a single GPU case, the following is the log file for training at 40 epoch:

```
2016-12-14 11:47:11,045 Node[0] Saved checkpoint to "cifar_resnet8-0-0040.params"
2016-12-14 11:47:13,342 Node[0] Epoch[39] Validation-cross-entropy=0.454520
2016-12-14 11:47:13,342 Node[0] Epoch[39] Validation-accuracy=0.844852
2016-12-14 11:47:16,796 Node[0] Epoch[40] Batch [50]	Speed: 1878.49 samples/sec	Train-cross-entropy=0.410888
2016-12-14 11:47:16,811 Node[0] Epoch[40] Batch [50]	Speed: 1878.49 samples/sec	Train-accuracy=0.853281
```

And the following is with `--load-epoch=40`:

```
2016-12-14 14:40:12,657 Node[0] Epoch[39] Validation-cross-entropy=0.454520
2016-12-14 14:40:12,657 Node[0] Epoch[39] Validation-accuracy=0.844852
2016-12-14 14:40:18,674 Node[0] Epoch[40] Batch [50]	Speed: 1828.57 samples/sec	Train-cross-entropy=0.401447
2016-12-14 14:40:18,674 Node[0] Epoch[40] Batch [50]	Speed: 1828.57 samples/sec	Train-accuracy=0.865156
```

**As can be seen, Validation-cross-entropy=0.454520 and Validation-accuracy=0.844852, they perfectly match for single GPU.**

However, when I am training the same code with 4 GPUs, the original log file is:

```
2016-12-14 13:52:55,005 Node[0] Epoch[39] Validation-cross-entropy=0.470790
2016-12-14 13:52:55,005 Node[0] Epoch[39] Validation-accuracy=0.838041
2016-12-14 13:52:57,130 Node[0] Epoch[40] Batch [50]	Speed: 3079.88 samples/sec	Train-cross-entropy=0.428925
2016-12-14 13:52:57,130 Node[0] Epoch[40] Batch [50]	Speed: 3079.88 samples/sec	Train-accuracy=0.856406
```

And the following is with `--load-epoch=40`:

```
2016-12-14 14:24:25,421 Node[0] Epoch[39] Validation-cross-entropy=0.443039
2016-12-14 14:24:25,421 Node[0] Epoch[39] Validation-accuracy=0.847756
2016-12-14 14:24:30,765 Node[0] Epoch[40] Batch [50]	Speed: 1896.30 samples/sec	Train-cross-entropy=0.388737
2016-12-14 14:24:30,765 Node[0] Epoch[40] Batch [50]	Speed: 1896.30 samples/sec	Train-accuracy=0.864531
```

**As can be seen, however, for  Validation-accuracy, 0.847756 (resuming from a checkpoint) is larger than 0.838041 (original) around 1% for multiple gpus.**

This phenomenon (when resuming from a multi-gpus training, the validation accuracy gets a little higher (around 1%) than the original one) appears multiple times when I am training the neural network in other cases. Is this a bug in the Multi GPU synchronization process?

**Comment by Trangle** [comment]
Try lr multiply 0.1 then run again.

**Comment by taoari** [comment]
@Trangle I made a modification to the code, evaluating the test dataset before training (see https://github.com/taoari/mxnet/blob/finetune-api/python/mxnet/model.py#L230_L254). 

So the evaluation accuracy is calculated right after loading the parameters, and is before the continuing training process. Hence multiply lr with 0.1 will have no effect on the results.

**Comment by mli** [comment]
multi-gpus should report (almost) the same accuracy as a simple gpu. one way to check is try https://github.com/dmlc/mxnet/blob/master/example/image-classification/score.py , it can load a model and evaluate it using various number of gpus.

**Comment by taoari** [comment]
@mli **Let me make the problem more clear**. I first trained ResNet-8 on Cifar10 dataset with **4 GPUs**, and achieved 0.838041 validation accuracy (ACC1) for the 40-th epoch. Then I resumed the checkpoint at 40-th epoch with **4 GPUs**, the evaluation accuracy becomes 0.847756 (ACC2). I also tried to resume the checkpoint at 40-th epoch with **single GPU**, the evaluation accuracy is also 0.847756 (ACC3).

Since ACC2 == ACC3 > ACC1, it is quite strange where the bug comes from. As ACC2 == ACC3, it indicates the multi gpu synchronization process is correct. But ACC2 > ACC1, it also indicates the multi gpu synchronization process might be wrong. This is quite confusing. 

P.S. Note that I added an initial evaluation of the checkpointed model performance right after loading the parameters, which is before resuming training. https://github.com/taoari/mxnet/blob/finetune-api/python/mxnet/model.py#L230_L254 is simply a copy of the evaluation code of https://github.com/taoari/mxnet/blob/finetune-api/python/mxnet/model.py#L329_L350. ACC2 == ACC3 > ACC1 shows that https://github.com/taoari/mxnet/blob/finetune-api/python/mxnet/model.py#L230_L254 is OK for both single GPU and multi GPUs, but https://github.com/taoari/mxnet/blob/finetune-api/python/mxnet/model.py#L329_L350 is not correct for multi GPUs. But these two parts are of exactly the same codes.

**Comment by zhreshold** [comment]
Just a vague guess, maybe in multi-GPU case the evaluation is conducted before the parameters are synchronized and updated? Cause the accuracy is always better with a small margin after resume loading the model. In single gpu case no synchronization is required.

**Comment by mli** [comment]
agree. probably `eval_data.reset()` is not correctly scheduled. might also related to #4253

@piiswrong

**Comment by piiswrong** [comment]
@mli One possible reason is saving get_param will reduce weight from multiple gpus. while during training they are not reduced. Is it possible for numerical error to accumulate during training?

**Comment by piiswrong** [comment]
@taoari Please check if this fixes it: https://github.com/dmlc/mxnet/pull/4327

**Comment by taoari** [comment]
@piiswrong Yes. The problem is fixed. I notice that you only modified the Module API, will you also make the same changes to the Model API?

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
